dbft_test.go

```diff
@@ -451,6 +451,230 @@ func TestDBFT_Invalid(t *testing.T) {
 	})
 }
 
+// TestDBFT_FourGoodNodesDeadlock checks that the following liveness lock is not really
+// a liveness lock and there's a way to accept block in this situation.
+// 0 :> [type |-> "cv", view |-> 1]           <--- this is the primary at view 1
+// 1 :> [type |-> "cv", view |-> 1]           <--- this is the primary at view 0
+// 2 :> [type |-> "commitSent", view |-> 0]
+// 3 :> [type |-> "commitSent", view |-> 1]
+//
+// Test structure note: the test is organized to reproduce the liveness lock scenario
+// described in https://github.com/neo-project/neo-modules/issues/792#issue-1609058923
+// at the section named "1. Liveness lock with four non-faulty nodes". However, some
+// steps are rearranged so that it's possible to reach the target network state described
+// above. It is done because dbft implementation contains additional constraints comparing
+// to the TLA+ model. See the steps
+func TestDBFT_FourGoodNodesDeadlock(t *testing.T) {
+	r0 := newTestState(0, 4)
+	r0.currHeight = 4
```

Comment:
First of all, it should be noted that in our tla+ specifications the speaker's index in the next consensus round of the same view is chosen as `(oldView + 1) % N` (so for block 0 it's firstly validator 0, then validator 1 and etc.). In our dBFT implementation it is chosen as `(int(c.BlockIndex) - int(viewNumber)) % len(c.Validators)` (surprise for me!), so for abstract block 0 primary is firstly validator 0, then validator 3 and etc.

It doesn't change the things, but should I fix it in the tla+ specification?

Reply:
It's OK to have a simpler spec wrt this, the main things are:
 * there is one primary
 * everyone agrees on what node that is
 * primary is changed on view change and block change
 * primaries are rotated in a cycle and treated equally, every node can be a primary

Both ways to select satisfy.